File: ex10_c_dev_kit/examples/select_command_example.c
/*****************************************************************************
 *                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      *
 *                                                                           *
 * This source code is the property of Impinj, Inc. Your use of this source  *
 * code in whole or in part is subject to your applicable license terms      *
 * from Impinj.                                                              *
 * Contact support@impinj.com for a copy of the applicable Impinj license    *
 * terms.                                                                    *
 *                                                                           *
 * (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               *
 *                                                                           *
 *****************************************************************************/

#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "ex10_api/application_registers.h"
#include "ex10_api/board_init.h"
#include "ex10_api/event_packet_parser.h"
#include "ex10_api/ex10_helpers.h"
#include "ex10_api/ex10_reader.h"
#include "ex10_api/gen2_tx_command_manager.h"
#include "ex10_api/rf_mode_definitions.h"
#include "ex10_api/trace.h"


/* Settings used when running this example */
static uint32_t const inventory_duration_ms = 500;  // Duration in milliseconds
static uint8_t const  antenna               = 1;
static uint16_t const rf_mode               = mode_148;
static uint16_t const transmit_power_cdbm   = 3000;
static uint8_t const  initial_q             = 4;
static bool const     dual_target           = true;
static uint8_t const  session               = 0;
static bool const     halt_on_all_tags      = false;
static bool const     tag_focus_enable      = false;
static bool const     fast_id_enable        = false;


/* Global state */
static struct Ex10Interfaces  ex10_iface  = {0u};
static struct InfoFromPackets packet_info = {0u, 0u, 0u, 0u, {0u}};

/* Gen2 parameters for this example */
/* Note for the mask: Everything is loaded in left-justified.
 * This means the bytes are loaded in via lsb, but the bits
 * are loaded in left to right as well, making them bitwise msb.
 * This was done for reader clarity to make it look like a
 * bitstream
 * EX:
 * load_mask = 0xa2f5
 * load_bits = 12
 * bits_loaded = 0xa2f
 *
 * load_mask = 0x10
 * load_bits = 4
 * bits_loaded = 0b0001
 */
static uint8_t        select_mask_buffer[2u] = {0u};
static struct BitSpan select_mask            = {select_mask_buffer, 12u};

static struct SelectCommandArgs select_args_001 = {
    .target      = SelectedFlag,
    .action      = Action001,
    .memory_bank = SelectEPC,
    .bit_pointer = 0,
    .bit_count   = 12,
    .mask        = &select_mask,
    .truncate    = false,
};

static struct SelectCommandArgs select_args_101 = {
    .target      = SelectedFlag,
    .action      = Action101,
    .memory_bank = SelectEPC,
    .bit_pointer = 0,
    .bit_count   = 12,
    .mask        = &select_mask,
    .truncate    = false,
};

static int run_inventory_rounds(enum SelectType const select_type)
{
    struct InventoryRoundControlFields inventory_config = {
        .initial_q            = initial_q,
        .max_q                = initial_q,
        .min_q                = initial_q,
        .num_min_q_cycles     = 1,
        .fixed_q_mode         = true,
        .q_increase_use_query = false,
        .q_decrease_use_query = false,
        .session              = session,
        .select               = select_type,
        .target               = 0,
        .halt_on_all_tags     = halt_on_all_tags,
        .tag_focus_enable     = tag_focus_enable,
        .fast_id_enable       = fast_id_enable,
    };

    struct InventoryRoundControl_2Fields inventory_config_2 = {
        .max_queries_since_valid_epc = 0};

    struct InventoryHelperParams ihp = {
        .antenna               = antenna,
        .rf_mode               = rf_mode,
        .tx_power_cdbm         = transmit_power_cdbm,
        .inventory_config      = &inventory_config,
        .inventory_config_2    = &inventory_config_2,
        .send_selects          = (select_type == SelectAll) ? false : true,
        .remain_on             = false,
        .dual_target           = dual_target,
        .inventory_duration_ms = inventory_duration_ms,
        .packet_info           = &packet_info,
        .verbose               = true,
    };

    if (ex10_iface.helpers->simple_inventory(&ihp))
    {
        return -1;
    };

    return 0;
}

static int select_command_example(void)
{
    // NOTE: This example assumes a single tag is used
    printf("Starting Select command example\n");

    printf("\nInventory with Sel=ALL, no Select command\n");
    run_inventory_rounds(SelectAll);
    printf("Done\n");

    uint16_t crc_of_interest = 0;
    if (packet_info.total_singulations)
    {
        crc_of_interest = packet_info.access_tag.stored_crc;
        printf("The Select mask will be 0x%02x - the CRC of tag with EPC=0x",
               ex10_iface.helpers->swap_bytes(crc_of_interest));
        for (size_t idx = 0u; idx < packet_info.access_tag.epc_length; idx++)
        {
            printf("%02x", packet_info.access_tag.epc[idx]);
        }
    }
    else
    {
        fprintf(stderr, "No tags found in inventory\n");
        return -1;
    }

    // Change the mask based on the tag recieved
    select_mask.data[0u] = (uint8_t)(crc_of_interest);
    select_mask.data[1u] = (uint8_t)(crc_of_interest >> 8u);
    select_args_001.mask = &select_mask;
    select_args_101.mask = &select_mask;

    // Create select commands
    struct Gen2CommandSpec select_cmd = {
        .command = Gen2Select,
        .args    = &select_args_001,
    };
    struct Gen2CommandSpec select_cmd2 = {
        .command = Gen2Select,
        .args    = &select_args_101,
    };

    // Add our new selects
    struct Ex10Gen2TxCommandManager const* g2tcm =
        get_ex10_gen2_tx_command_manager();

    struct Gen2TxCommandManagerError error_1 =
        g2tcm->encode_and_append_command(&select_cmd, 0);
    assert(false == error_1.error_occurred);
    struct Gen2TxCommandManagerError error_2 =
        g2tcm->encode_and_append_command(&select_cmd2, 1);
    assert(false == error_2.error_occurred);
    // set the buffer to the device
    struct Gen2TxCommandManagerError error_3 = g2tcm->write_sequence();
    assert(false == error_3.error_occurred);

    // enable the first select command
    bool select_enables[MaxTxCommandCount] = {0u};
    select_enables[error_1.current_index]  = true;
    g2tcm->write_select_enables(select_enables, MaxTxCommandCount);

    printf("\nInventory with Sel=SL, sending Select with Action001\n");
    run_inventory_rounds(SelectAsserted);

    if (packet_info.total_singulations == 0u)
    {
        fprintf(stderr, "Selected tag did not reply\n");
        return -1;
    }
    else
    {
        printf("Done\n");
    }

    // enable the second select
    select_enables[error_1.current_index] = false;
    select_enables[error_2.current_index] = true;
    g2tcm->write_select_enables(select_enables, MaxTxCommandCount);

    printf("\nInventory with Sel=~SL, sending Select with Action101\n");
    run_inventory_rounds(SelectNotAsserted);

    if (packet_info.total_singulations == 0u)
    {
        fprintf(stderr, "Selected tag did not reply\n");
        return -1;
    }
    else
    {
        printf("Done\n");
    }

    printf("\nInventory with Sel=SL, sending Select with Action101\n");
    run_inventory_rounds(SelectAsserted);

    if (packet_info.total_singulations > 0u &&
        crc_of_interest == packet_info.access_tag.stored_crc)
    {
        fprintf(stderr, "Tag replied when was not selected\n");
        return -1;
    }
    else
    {
        printf("Done\n");
    }

    printf("Ending Select command example\n");

    return 0;
}

int main(void)
{
    tracepoint(pi_ex10sdk, EXEC_start, __FILE__);
    ex10_iface = ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, "FCC");

    int result = ex10_iface.helpers->check_board_init_status(Application);
    if (result == 0)
    {
        result = select_command_example();
    }

    ex10_typical_board_teardown();
    tracepoint(pi_ex10sdk, EXEC_end, __FILE__);

    return result;
}

File: ex10_dev_kit/examples/select_command_example.py
#############################################################################
#                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      #
#                                                                           #
# This source code is the property of Impinj, Inc. Your use of this source  #
# code in whole or in part is subject to your applicable license terms      #
# from Impinj.                                                              #
# Contact support@impinj.com for a copy of the applicable Impinj license    #
# terms.                                                                    #
#                                                                           #
# (c) Copyright 2019 - 2021 Impinj, Inc. All rights reserved.               #
#                                                                           #
#############################################################################
"""
The script below is an example to show how to send a Gen2 Select command using
the Ex10 reader API.
"""

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)

import binascii
from datetime import datetime

from py2c_interface.py2c_python_wrapper import *


# The configuration of the inventory example is set using the definitions below
INVENTORY_DURATION_S = 1            # Duration of inventory operation (seconds)
TRANSMIT_POWER_DBM = 3000           # 30 dBm
RF_MODE = RfModes.mode_5
REGION = 'FCC'                      # Regulatory region
R807_ANTENNA_PORT = 1               # Which R807 antenna port will be used
INITIAL_Q = 4                       # Q field in the Query command
SELECT_ALL = 0                      # SELECT field in the Query command
SELECT_NOT_ASSERT = 2
SELECT_ASSERT = 3
SESSION = 0                         # Session field in the Query command
TARGET = 0                          # Target field in the Query command
DUAL_TARGET = True                  # If True, the Target of subsequent Query
                                    # commands will flip

pc = c_uint16()
epc = (c_ubyte * 12)()
crc = c_uint16()

packet_info = InfoFromPackets(0, 0, 0,
                              TagReadData(pointer(pc),
                                          ctypes.cast(epc, POINTER(c_ubyte)),
                                          0,
                                          pointer(crc),
                                          None,
                                          0))


def run_inventory_rounds(ex10_reader, helper, select_type):
    """
    Runs an inventory round.
    :param ex10_reader: The reader object
    :param select_type: The type of select to configure the
                        inventory round using.
    """
    # Put together configurations for the inventory round
    inventory_config = InventoryRoundControlFields()
    inventory_config.initial_q = INITIAL_Q
    inventory_config.max_q = INITIAL_Q
    inventory_config.min_q = INITIAL_Q
    inventory_config.num_min_q_cycles = 1
    inventory_config.fixed_q_mode = True
    inventory_config.q_increase_use_query = False
    inventory_config.q_decrease_use_query = False
    inventory_config.session = SESSION
    inventory_config.select = select_type
    inventory_config.target = TARGET
    inventory_config.halt_on_all_tags = False
    inventory_config.fast_id_enable = False
    inventory_config.tag_focus_enable = False

    inventory_config_2 = InventoryRoundControl_2Fields()
    inventory_config_2.max_queries_since_valid_epc = 0

    ihp = InventoryHelperParams()
    ihp.antenna               = R807_ANTENNA_PORT
    ihp.rf_mode               = RF_MODE
    ihp.tx_power_dbm          = TRANSMIT_POWER_DBM
    ihp.inventory_config      = pointer(inventory_config)
    ihp.inventory_config_2    = pointer(inventory_config_2)
    ihp.send_selects          = False if select_type == SELECT_ALL else True
    ihp.remain_on             = False
    ihp.dual_target           = DUAL_TARGET
    ihp.inventory_duration_ms = INVENTORY_DURATION_S * 1000
    ihp.packet_info           = pointer(packet_info)
    ihp.enforce_gen2_response = False
    ihp.verbose               = True

    assert 0 == helper.simple_inventory(ihp)


def load_select_commands(tx_command_manager, selected_tag_crc, crc_byte_len):
    """
    Loads select commands into the Gen2 Buffer
    :param selected_tag_crc: The crc to use in building
                             the select command.
    :param crc_byte_len: The length of the crc to use as the mask
    """
    print('Loading commands into Gen2 command buffer')

    crc_bytes = bytearray.fromhex('{:04x}'.format(selected_tag_crc))
    crc_bytes.reverse()
    arraytype = ctypes.c_uint8 * crc_byte_len
    mask_arr = arraytype()
    byte_arr = BitSpan(pointer(c_uint8(addressof(mask_arr))), int(crc_byte_len)*8)
    for ind in range(crc_byte_len):
        byte_arr.data[ind] = crc_bytes[ind]

    select_args = SelectCommandArgs()
    select_args.target = SelectTarget.SelectedFlag
    select_args.action = SelectAction.Action101
    select_args.memory_bank = SelectMemoryBank.SelectEPC
    select_args.bit_pointer = 0
    select_args.truncate = 0
    select_args.mask = pointer(byte_arr)
    select_args.bit_count = 8*crc_byte_len

    cmd = Gen2CommandSpec(Gen2Command.Gen2Select,
                          ctypes.cast(pointer(select_args), c_void_p))

    # Clear any commands before adding new ones
    tx_command_manager.clear_local_sequence();
    enables = [0]*10
    # Add the command
    curr_error = tx_command_manager.encode_and_append_command(cmd, 0);
    assert(False == curr_error.error_occurred);
    enables[curr_error.current_index] = True;

    # Change params for a 001 action
    select_args.action = SelectAction.Action001
    # Add as as new command
    curr_error = tx_command_manager.encode_and_append_command(cmd, 1);
    assert(False == curr_error.error_occurred);
    enables[curr_error.current_index] = True;
    curr_error = tx_command_manager.write_sequence();
    assert(False == curr_error.error_occurred);


def enable_select_command(tx_command_manager, enable_index):
    """
    Enable one select at a time for this scenario via the index.
    :param enable_index: The index of the select to enable
    """
    arraytype = ctypes.c_bool * 10
    enable_me = arraytype(0,0,0,0,0,0,0,0,0,0)
    enable_me[enable_index] = 1
    enable_p = pointer(enable_me)

    tx_command_manager.write_select_enables(ctypes.cast(enable_p, c_void_p), 10);


def run_select_script(ex10_ifaces, tx_command_manager):
    """ Run inventory for the specified amount of time """
    ex10_reader = ex10_ifaces.reader
    ex10_ops = ex10_ifaces.ops
    helper = ex10_ifaces.helpers

    print('Starting Select example')
    try:
        # Perform an inventory round to find a tag that we can
        print('Inventory with Sel=ALL, no Select command')
        run_inventory_rounds(ex10_reader, helper, SELECT_ALL)
        if 0 == packet_info.total_singulations:
            raise RuntimeError("No tags found in basic inventory")
        print('Done')

        # We will use the CRC of the EPC as mask for the select commands
        crc_hex_string = '{:04x}'.format(packet_info.access_tag.crc[0])
        crc_byte_len = int(len(crc_hex_string)/2) # number of nibbles in the string / 2
        tag_epc_list = packet_info.access_tag.epc[:packet_info.access_tag.epc_length]
        tag_epc_hex_list = ('{:02x}'.format(i) for i in tag_epc_list)
        tag_epc_hex_string = ''.join(map(str, tag_epc_hex_list))
        print('The Select mask will be 0x{} (the CRC of tag with EPC=0x{})'.format(
            crc_hex_string, tag_epc_hex_string))

        # Load in select commands based off the received tag crc.
        load_select_commands(tx_command_manager, packet_info.access_tag.crc[0], crc_byte_len)

        # The command to select on SL is in the buffer under index 1. Enable
        # to run before inventory.
        enable_select_command(tx_command_manager, 1)

        # Run Inventory with SEL=SL(11) and a Select command that asserts SL
        # and check that only the selected tag replies
        print('Inventory with Sel=SL, sending Select with Action001')
        run_inventory_rounds(ex10_reader, helper, SELECT_ASSERT)
        if 0 == packet_info.total_singulations:
            raise RuntimeError("Selected tag did not reply")

        print('Done')

        # The command to select on !SL is in the buffer under index 0. Enable
        # to run before inventory.
        enable_select_command(tx_command_manager, 0)

        # Run Inventory with SEL=~SL(10) and a Select command that deasserts SL
        # and check that selected tag replied
        print('Inventory with Sel=~SL, sending Select with Action101')
        run_inventory_rounds(ex10_reader, helper, SELECT_NOT_ASSERT)
        if 0 == packet_info.total_singulations:
            raise RuntimeError("Selected tag did not reply")

        print('Done')

        # The command to select on !SL is in the buffer under index 0 which
        # is already enabled.

        print('Inventory with Sel=SL, sending Select with Action101')
        run_inventory_rounds(ex10_reader, helper, SELECT_ASSERT)
        if 0 != packet_info.total_singulations:
            tag_epc_list_new = packet_info.access_tag.epc[:packet_info.access_tag.epc_length]
            if tag_epc_list == tag_epc_list_new:
                raise RuntimeError("Tag replied when was not selected")

        print('Done')

    finally:
        ex10_reader.stop_transmitting()
        print('Select example end')


def run_select_example():
    """ Run example using select command and Gen2 Tx buffer """
    try:
        # Init the python to C layer
        py2c = Ex10Py2CWrapper()
        ex10_ifaces = py2c.ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, 'FCC'.encode('ascii'))
        tx_command_manager = py2c.get_ex10_gen2_tx_command_manager()

        # Run the Select example
        run_select_script(ex10_ifaces, tx_command_manager)
    finally:
        py2c.ex10_typical_board_teardown()


if __name__ == "__main__":
    run_select_example()
